### PR TITLE
Step 1212 pathutil

### PR DESCRIPTION
**Pull request by godrei**
### Context

This PR introduces `PathProvider`, `PathChecker` and `PathModifier` to support Xcode Test Step's ideal step refactor.

### Changes

- Introduce `PathModifier` to break direct dependency to AbsPath function
- Introduce `PathChecker` to break direct dependency to IsPathExists function

Non-functional changes:

- Rename `TempDirProvider` to `PathProvider` to support adding more related functions later
- `EscapeGlobPath` function moved to pathutil.go to keep function related to file path modification in the same file
- `ListEntries` function moved to path_filter.go as it provides path listing with FilterFunc
- `ListPathInDirSortedByComponents` function moved to sortable_path.go as it provides sorted path listing
- function in pathutil.go grouped into `Path provider functions`, `Path checker functions`, `Path modifier functions` and `Change dir functions`

### CODE REVIEW

File: pathutil/pathutil.go [hunk 1]
```diff
@@ -10,64 +10,54 @@ import (
...
+//
+// Path provider functions
+
+// PathProvider ...
+type PathProvider interface {
+	TempDir(prefix string) (string, error)
 }
 
-type defaultTempDirProvider struct{}
+type pathProvider struct{}
```
Comment (Bence1001) [inline pathutil/pathutil.go:21]: This should be `defaultPathProvider`.

File: pathutil/pathutil.go [hunk 2]
```diff
@@ -10,64 +10,54 @@ import (
...
 
-// TempDirProvider ...
-type TempDirProvider interface {
-	CreateTempDir(prefix string) (string, error)
+//
+// Path provider functions
+
+// PathProvider ...
+type PathProvider interface {
+	TempDir(prefix string) (string, error)
```
Comment (Bence1001) [inline pathutil/pathutil.go:18]: If this creates a directory, shouldn't we keep the `Create` prefix in the function name?

File: pathutil/pathutil.go [hunk 3]
```diff
@@ -79,6 +69,24 @@ func EnsureDirExist(dir string) error {
 	return nil
 }
 
+//
+// Path checker functions
+
+// PathChecker ...
+type PathChecker interface {
+}
```
Comment (Bence1001) [inline pathutil/pathutil.go]: Empty interface. Is this intentional?

File: pathutil/pathutil.go [hunk 4]
```diff
@@ -79,6 +69,24 @@ func EnsureDirExist(dir string) error {
...
+
+type pathChecker struct{}
```
Comment (Bence1001) [inline pathutil/pathutil.go:80]: This should be called `defaultPathChecker`.

File: pathutil/pathutil.go [hunk 5]
```diff
@@ -123,6 +125,32 @@ func IsDirExists(pth string) (bool, error) {
...
+
+//
+// Path modifier functions
+
+// PathModifier ...
+type PathModifier interface {
+	AbsPath(pth string) (string, error)
+}
+
+type pathModifier struct{}
```
Comment (Bence1001) [inline pathutil/pathutil.go:143]: This should be called `defaultPathModifier`.

File: pathutil/pathutil.go [hunk 6]
```diff
@@ -10,64 +10,54 @@ import (
...
 )
 
-// TempDirProvider ...
-type TempDirProvider interface {
-	CreateTempDir(prefix string) (string, error)
+//
+// Path provider functions
+
+// PathProvider ...
+type PathProvider interface {
```
Comment (Bence1001) [inline pathutil/pathutil.go:17]: You've mentioned this has been renamed to allow adding more functions in the future that are not necessarily related to temp dirs. What are those use cases?

**Review by Bence1001 — CHANGES_REQUESTED** [review]

File: pathutil/pathutil.go (hunk 6 above)
Comment (godrei) [reply pathutil/pathutil.go:17]: The functions between `// Path provider functions` and `// Path checker functions` comments:
- `NormalizedOSTempDirPath` (already added as `TempDir`)
- `CurrentWorkingDirectoryAbsolutePath`
- `UserHomeDir`
- `EnsureDirExist`

I feel like these functions are used to get some sort of file path.

File: pathutil/pathutil.go (hunk 3 above)
Comment (godrei) [reply pathutil/pathutil.go]: that was unintentional, fixed

File: pathutil/pathutil.go (hunk 6 above)
Comment (Bence1001) [reply pathutil/pathutil.go:17]: Let's move forward with this naming scheme then for now. 👍

**Review by Bence1001 — APPROVED** [review]

**Pull request merged**